### DISCUSSION TITLE
Question about the syntax for insert with select in values

### DISCUSSION THREAD

**Discussion by mrclrchtr** [jOOQ/jOOQ#16873] [Q&A] [answered]
Maybe it's a dumb question, but I've been looking for the solution for a while now, so I'm asking here:

This is working:
```kotlin
fun working(
    dsl: DSLContext,
    tenantId: Int,
    contactId: Int,
  ): Int {
    return dsl.insertInto(CONTACT_IDENTIFIER_CUSTOMER)
      .values(
        contactId,
        tenantId,
        dsl
          .select(
            coalesce(max(CONTACT_IDENTIFIER_CUSTOMER.CUSTOMER_ID), 0).add(1)
          )
          .from(CONTACT_IDENTIFIER_CUSTOMER)
          .where(
            CONTACT_IDENTIFIER_CUSTOMER.TENANT_ID.eq(tenantId)
          )
      )
      .execute()
  }
```

but when I want t specify the order of the values to insert like this:

```kotlin
  fun syntax(
    dsl: DSLContext,
    tenantId: Int,
    contactId: Int,
  ): Int {
    return dsl.insertInto(
      CONTACT_IDENTIFIER_CUSTOMER,
      CONTACT_IDENTIFIER_CUSTOMER.CONTACT_ID,
      CONTACT_IDENTIFIER_CUSTOMER.TENANT_ID,
      CONTACT_IDENTIFIER_CUSTOMER.CUSTOMER_ID
    ).values(
      contactId,
      tenantId,
      dsl
        .select(
          coalesce(max(CONTACT_IDENTIFIER_CUSTOMER.CUSTOMER_ID), 0).add(1)
        )
        .from(CONTACT_IDENTIFIER_CUSTOMER)
        .where(
          CONTACT_IDENTIFIER_CUSTOMER.TENANT_ID.eq(tenantId)
        )
    ).execute()
  }
```

I get the following syntax error:

```
None of the following functions can be called with the arguments supplied.
values(Int?, Int?, Int?) defined in org.jooq.InsertValuesStep3
values(Field<Int?>!, Field<Int?>!, Field<Int?>!) defined in org.jooq.InsertValuesStep3
```

My intention is to make sure that the values are in the right place. I had a bug before because I accidentally swapped `contactId` and `tenantId` in the values.

Thank you very much!

**Comment by lukaseder** [comment, marked as answer]
> This is working:

Yes, it works, because it lacks type safety, so jOOQ will just coerce the `VALUES()` arguments to the most reasonable thing at runtime. But you shouldn't use this syntax, just like you shouldn't use `SELECT *`,  because the behaviour will depend on the DDL column definition order, which may be different between your database installations (test and production), depending on how you created or altered the table.

> but when I want t specify the order of the values to insert like this:

Just wrap the `Select` in a `Field` using `DSL.field(Select)`. This is called a "scalar subquery":
https://www.jooq.org/doc/latest/manual/sql-building/column-expressions/scalar-subqueries/

Also, you cannot mix `T` and `Field<T>` types in a `VALUES` list, so wrap all bind values in `DSL.value()` as well:
- https://stackoverflow.com/q/67522537/521799
- https://www.jooq.org/doc/latest/manual/sql-building/bind-values/indexed-parameters/

**Reply by mrclrchtr** [reply]
Thank you so much! Especially for the additional note about `val`!
